internal/util/merge/merge3_test.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merge
+
+import (
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/internal/testutil/pkgbuilder"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/kustomize/kyaml/copyutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+func TestMerge3_Nested_packages(t *testing.T) {
+	annotationSetter := yaml.SetAnnotation("foo", "bar")
+	labelSetter := yaml.SetLabel("bar", "foo")
+
+	testCases := []struct {
+		name               string
+		includeSubPackages bool
+		original           *pkgbuilder.Pkg
+		upstream           *pkgbuilder.Pkg
+		local              *pkgbuilder.Pkg
+		expected           *pkgbuilder.Pkg
+	}{
+		{
+			name:               "subpackages are merged if included",
+			includeSubPackages: true,
+			original:           createPkg(),
+			upstream:           createPkg(annotationSetter),
+			local:              createPkg(labelSetter),
+			expected:           createPkg(labelSetter, annotationSetter),
+		},
+		{
+			name:               "subpackages are not merged if not included",
+			includeSubPackages: false,
+			original:           createPkg(),
+			upstream:           createPkg(annotationSetter),
+			local:              createPkg(labelSetter),
+			expected: createPkgMultipleMutators(
+				[]yaml.Filter{
+					labelSetter,
+					annotationSetter,
+				},
+				[]yaml.Filter{
+					labelSetter,
+				},
+			),
+		},
+		{
+			name:               "local copy defines the package boundaries if different from upstream",
```

Comment:
These are very remote scenarios which user might never run into. I am not sure we should go this deep in order to make the update package aware. All these scenarios can still be addressed by merging files instead of packages. It is easier to message user about the expected behavior as well. Having multitude of scenarios makes it tougher to convey the expected behavior.